### ISSUE TITLE
compile with new versions of Lazarus(2.0.10) and fpc causes troubles

### ISSUE THREAD

**Issue by OH1KH** [ok2cqr/cqrlog#323]
As a warning sign:
After installed the latest Lazarus (and fpc) some changes can be found:
non fatal:
- TRXControl: if rig not conneted used to show "0.00000" now shows "0" as frequency

Fatal:
- regular expressions do not allow empty input strings. Causes runtime warnings.
  For now on empty string should be checked before passing to regex routines.

I a'm now working to find all places where they are used (many, but how many....)
  - At least qsl label printing fails with qsl viacall validate.
  - HamQTH info fetch fails with DOK info if user has no DOK set

 I am afraid there is more to come...

**Comment by OH1KH** [comment]
Reverted back to Laz2.0.8/fpc3.0.4. It works.

Additional fatal problem is UTF8

When fixed regexp mentioned before I tested with empty new log and callsign E77A. His qth shows out ok when fetch from HamQTH but when logged the first letter of qth (S with ~ on top) is something else ( Å iroki Brijeg) in QSO list.
It is so also in database and qsl label export.

But when I compile with 2.0.8/3.0.4 QSO list view shows ok, View QSO loads ok and qsl label export exports ok.
Command line console mysql still shows  Å iroki Brijeg like with Laz2.0.10

I think I stay in 2.0.8 (forever?)

**Comment by OH1KH** [comment]
Found this just from Laz forum
https://github.com/andgineer/TRegExpr/issues/137

**Comment by OH1KH** [comment]
Today I installed again Laz2.0.10/fpc3.2.0 to see if regexp fixes done had effect. So far I did not see any regexp error splash but I did not test very long time because I wanted to see is  there still and UTF-8 problem.

Yes ! there is !
**Until this is solved DO NOT USE LAZ 2.0.10/fpc3.2.0  for making cqrlog !!!**
It may be a need of compile directive. that would be easy. Or bigger trouble.

![utf_error](https://user-images.githubusercontent.com/12532164/107488076-8f15d480-6b8f-11eb-9a03-73eebc15d757.png)

**Comment by OH1KH** [comment]
Continued testing:
"Garbage" output happens also in ADIF and HTML exports.
ADIF exports properly switches tag from <REMARKS to <REMARKS_INTL when special character are found, but tag content results are not readable.

to be continued....

**Comment by OH1KH** [comment]
Continued testing:
Same problem when qso is imported from fldigi or wsjt-x remote. Special characters there will show up as "garbage" in cqrlog's View qso or export adif and html.

Seems that placing text to column's TEdit.Text from anywhere else than PC's keyboard results false special characters.

to be continued....

**Comment by OH1KH** [comment]
It looks like this change in fpc 3.2.0 would have that kind of change. How ever I can not understand how that reflects to TEdit.Text 

Classes TStrings.LoadFromStream/File encoding handling
Old behaviour: The LoadFromStream call totally ignored the encoding of a file, loading it from a stream without regard for encoding, unless an encoding was specified.
New behaviour: There is an overloaded call with a Boolean parameter 'IgnoreEncoding' which determines whether the encoding should be taken into account or not. By default, this parameter is False, meaning that LoadFromStream with just a stream as argument now calls the encoding-aware version of LoadFromStream, passing it an encoding of Nil, which mean the default encoding will be used.
Reason for change: Delphi compatibility, and the corresponding changes in TStringStream constructors.
Remedy: The old behaviour can be restored by setting the IgnoreEncoding parameter to 'True'.
svn: 37962 and 37965

It seems to be best to use Laz2.0.8/fpc3.0.4 combination for ever. That should be noted to all packet maintainers.

I think I go out and do some snow works instead. It is human understandable work ;-(

**Comment by OH1KH** [comment]
OK !
It seems to be this, how ever:
CodePage aware TStringField and TMemoField
Old behaviour: These character fields were agnostic to data they presented. IOW when we read content of such field using AsString data was passed from internal character buffer to string as is without any conversion.
New behaviour: When those fields are created there can be specified CodePage (if none specified CP_ACP is assumed), which defines encoding of character data presented by this field. In case of sqlDB TSQLConnection.CharSet is usualy used. When we read content of such field using AsString character data are translated from fields code page to CP_ACP. Same translation happens when we read content using AsUTF8String (translation to CP_UTF8) or AsUnicodeString (translation to UTF-16). This change reflects CodePage aware strings introduced in FPC 3.

Just how to fix source so that old and new fpc compilded  versions work crossed ok.

**Comment by OH1KH** [comment]
There is a way to get this working:

Before database connection, set character at  dmData.OpenConnections:
 MainCon.CharSet:='UTF8';
Do same for all 5 connections.

But this leads to situation that is not backward compatible. When you read and write qso records characters are ok.
But when you view old qsos characters are not ok any more.
Or if you open log with fpc304 compiled program then old qsos have ok characters but new fpc320 compiled program entries do not have special characters ok.

So this method can not be used. Unless we do database conversion routine that transfers old special characters to new ones. But then there is no coming back to fpc304 compiled programs.

**Comment by OH1KH** [comment]
This sql connection CharSet has been a long standing bug. It is just been hidden before fpc3.2.0.

Adding MainCon.CharSet:='UTF8'; (and other connects, too) to fpc3.0.4 compilation does same effect and actually makes everything work as should as then peek to database with Linux console shows special characters OK also there.

To get this in production simplest way for programmers and users is to make full export to adif with old program compiled by fpc3.0.4 and then update to program compiled with fpc3.2.0 and import adif.
That way all qsos with special characters in data gets converted to true UTF-8

How ever I have feeling that users may fail with this/ignore it and then after a while start complaining.

**Comment by OH1KH** [comment]
I will add one more comment to this thread.
Perhaps we could make database version upgrade routine. Same way as normally done when database structure changes.

Special characters mostly exist in table cqrlog_main. If just go through that reading all possible columns that could have special characters using read ".AsBytes".
Then use function I introduced in this thread earlier and push things back to columns they were read from.
This can be done with with fpc3.2.0 compiled version that uses MainCon.CharSet:='UTF8'; as columns are read AsBytes.

I think that is all for this.
Problem is defined, solution exist, now just implementation is needed.

**Comment by OH1KH** [comment]
(reply here as PR #399 is closed)
Petr encoding needs not to be utf8mb4 See below (I have not verified that by my self):

"According to the sources the spellings 'utf8', 'utf-8' and 'utf8mb4' for the charset property should have the same effect."
Whole story: https://forum.lazarus.freepascal.org/index.php/topic,53250.0.html

What I put into pull request is enough to get all working. If you change to utf8mb4 you have to change all database definitions too if you like to do it perfect.
Now when tables are utf8 it is enough to set connection charset properly.

Now when you have accepted pull request "as is" be aware that all users having special characters like Scandinavian characters, U umlaut or letters used in Spanish language will **not see previously logged qsos properly any more** as I already stated.

**ADIF export** must be done **before** new version is installed and **after** it is installed **ADIF import must be done to empty log**.

**Comment by OH1KH** [comment]
Now what we need after PR #399 is a conversion procedure that reads old log entries columns and converts them to new charset UTF8.

It is easiest apply in NewQSO as "File/Convert CharSet in old loggings"

Could be also in QSOList "Data/Convert CharSet in old loggings" , then fNewQSO.CreateAutobackup have to be moved to "public" from "private"

Selecting that creates a new form where:

	- User must set "date up to" until where the conversion is to be done.
	  (if after PR#339 qsos are converted it makes garbage for them )
	- checks that given date is right format
	- checks that user has set autobackup path and requires it to be set if not set.
	  After that enables "convert" button.
	- when "convert" button is pressed calls first fNewQSO.CreateAutobackup
	- after autobackup goes through all columns in all records form following tables using:

	  	- read.AsBytes to read column from database
		- uses function for all columns read AsBytes;
			function TdmUtils.DoString(BytesIn:Tbytes):string;
			var
			 s:String = '';
			 i: integer;
			Begin
			 for i:=0 to length(BytesIn) do
			 s:=s+char(BytesIn[i]);
			 Result :=s;
			end;
		- returns column string data back to database with sql UPDATE/Set/where


columns from cqrlog_main to be passed through:

| name           | varchar(40)   | YES  | MUL |         |                |
| qth            | varchar(60)   | YES  | MUL |         |                |
| county         | varchar(30)   | YES  |     |         |                |
| award          | varchar(50)   | YES  |     |         |                |
| remarks        | varchar(200)  | YES  |     |         |                |
| club_nr1       | varchar(100)  | YES  | MUL |         |                |
| club_nr2       | varchar(100)  | YES  | MUL |         |                |
| club_nr3       | varchar(100)  | YES  | MUL |         |                |
| club_nr4       | varchar(100)  | YES  | MUL |         |                |
| club_nr5       | varchar(100)  | YES  | MUL |         |                |
| satellite      | varchar(30)   | YES  |     |         |                |
| prop_mode      | varchar(30)   | YES  |     |         |                |
| stx_string     | varchar(50)   | YES  |     | NULL    |                |
| srx_string     | varchar(50)   | YES  |     | NULL    |                |
| contestname    | varchar(40)   | YES  |     | NULL    |                |
| operator       | varchar(20)   | YES  |     | NULL    |        


columns from notes to be passed through:
| longremarks | varchar(256) | YES  |     |         |                |

columns from long_note to be passed through:
| note         | longtext | YES  |     | NULL    |                |

columns from freqmem to be passed through:
| info      | varchar(25)   | YES  |     | NULL    |                |


This is easiest way for user to get things back to order.

Any possible other columns in any database?
Comments?
Any other ideas?

**Comment by OH1KH** [comment]
What happened?

After merged master with UTF8 bugfix and made new branch for planning some kind of fix routine for user to fix old database entries having special characters now showing out as garbage my method reading AsBytes and using convert function I introduced here  in last message **does not work any more!**

I have spent too many days for this UTF issue. I can live without proper ÖÄÅ in old qso entries.
This is not my problem any more.
